- Identify the data types you are working with.
- Examine the distributions of your data, numerically and/or visually.
- Identify outliers.
- Identify missing data and look for patterns of missing data.
- Describe how your EDA will inform your modeling decisions and process.


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import datetime as datetime


## Import data

In [63]:
tickers = pd.read_csv('../dataset/tickers.csv')
#name the header ticker

In [66]:
tickers

ticker
0      ABIO
1      AGTC
2       AIM
3      AIRT
4      ALSK
...     ...
1943   CAGZ
1944   PTSC
1945   JUHL
1946   OPST
1947   IDAH

[1948 rows x 1 columns]

In [68]:
tickers['ticker'].str.contains('WNEB').any()

True

### price_volume_sorted data

In [3]:
price_volume_sorted = pd.read_csv('../dataset/price_volume_sorted.csv',  parse_dates=['date'])
price_volume_sorted.head()

ticker       date   open  high    low  close   volume  dividends  \
0   AAME 2015-01-02  3.990  4.03  3.980   4.03  11443.0        0.0   
1   AAME 2015-01-05  3.900  4.01  3.900   4.01  13727.0        0.0   
2   AAME 2015-01-06  3.950  3.95  3.750   3.92   9743.0        0.0   
3   AAME 2015-01-07  3.899  3.92  3.870   3.92   1486.0        0.0   
4   AAME 2015-01-08  3.920  3.95  3.915   3.95   2200.0        0.0   

   closeunadj lastupdated  
0        4.03  2018-06-13  
1        4.01  2018-06-13  
2        3.92  2018-06-13  
3        3.92  2018-06-13  
4        3.95  2018-06-13

In [4]:
price_volume_sorted.dtypes

ticker                 object
date           datetime64[ns]
open                  float64
high                  float64
low                   float64
close                 float64
volume                float64
dividends             float64
closeunadj            float64
lastupdated            object
dtype: object

In [5]:
price_volume_sorted.isnull().sum()

ticker         0
date           0
open           0
high           0
low            0
close          0
volume         0
dividends      0
closeunadj     0
lastupdated    0
dtype: int64

In [6]:
price_volume_sorted.head()

ticker       date   open  high    low  close   volume  dividends  \
0   AAME 2015-01-02  3.990  4.03  3.980   4.03  11443.0        0.0   
1   AAME 2015-01-05  3.900  4.01  3.900   4.01  13727.0        0.0   
2   AAME 2015-01-06  3.950  3.95  3.750   3.92   9743.0        0.0   
3   AAME 2015-01-07  3.899  3.92  3.870   3.92   1486.0        0.0   
4   AAME 2015-01-08  3.920  3.95  3.915   3.95   2200.0        0.0   

   closeunadj lastupdated  
0        4.03  2018-06-13  
1        4.01  2018-06-13  
2        3.92  2018-06-13  
3        3.92  2018-06-13  
4        3.95  2018-06-13

In [7]:
price_volume_sorted.tail()

ticker       date   open   high     low  close    volume  dividends  \
2397143   ZYXI 2020-12-14  13.10  13.59  13.100  13.34  446678.0        0.0   
2397144   ZYXI 2020-12-15  13.45  13.45  13.030  13.18  372095.0        0.0   
2397145   ZYXI 2020-12-16  13.24  13.95  13.240  13.87  560581.0        0.0   
2397146   ZYXI 2020-12-17  13.94  14.29  13.701  14.25  409889.0        0.0   
2397147   ZYXI 2020-12-18  14.25  14.39  13.730  14.00  551332.0        0.0   

         closeunadj lastupdated  
2397143       13.34  2020-12-14  
2397144       13.18  2020-12-15  
2397145       13.87  2020-12-16  
2397146       14.25  2020-12-17  
2397147       14.00  2020-12-18

## Basic EDA and Feature Engineering

### Price
- Since we have four values of stock price for each day, let's create a feature called Price which is the average of all these values.

In [8]:
df = price_volume_sorted
#store it as df

In [9]:
values = (df['high'] + df['low'] + df['open'] + df['close'])/4
df = df.assign(Price=values)

In [10]:
df.head()


ticker       date   open  high    low  close   volume  dividends  \
0   AAME 2015-01-02  3.990  4.03  3.980   4.03  11443.0        0.0   
1   AAME 2015-01-05  3.900  4.01  3.900   4.01  13727.0        0.0   
2   AAME 2015-01-06  3.950  3.95  3.750   3.92   9743.0        0.0   
3   AAME 2015-01-07  3.899  3.92  3.870   3.92   1486.0        0.0   
4   AAME 2015-01-08  3.920  3.95  3.915   3.95   2200.0        0.0   

   closeunadj lastupdated    Price  
0        4.03  2018-06-13  4.00750  
1        4.01  2018-06-13  3.95500  
2        3.92  2018-06-13  3.89250  
3        3.92  2018-06-13  3.90225  
4        3.95  2018-06-13  3.93375

In [11]:
df.describe().apply(lambda s: s.apply('{0:.5f}'.format))

open              high               low             close  \
count     2397148.00000     2397148.00000     2397148.00000     2397148.00000   
mean        89203.14106       94416.44721       84003.94079       87851.82232   
std       8560189.38525     9156852.49674     8064928.76648     8443640.11737   
min             0.00001           0.00010           0.00001           0.00010   
25%             1.53000           1.59000           1.50000           1.53000   
50%             5.75000           5.91000           5.60000           5.75000   
75%            15.00000          15.23000          14.75000          15.00000   
max    1626238373.76500  2049597950.40600  1555678444.32500  1673278326.72500   

                volume      dividends     closeunadj             Price  
count    2397148.00000  2397148.00000  2397148.00000     2397148.00000  
mean      219906.91451        0.00516       42.32264       88868.83785  
std      2918537.65131        2.76311     2181.89773     8546078.74477  
min            0.00000        0.00000        0.00000           0.00006  
25%         1751.00000        0.00000        0.86000           1.53500  
50%        15782.00000        0.00000        3.56000           5.75000  
75%        76277.00000        0.00000       11.37000          15.00000  
max    955210864.00000     3080.00000   172250.00000  1724509875.49350

In [12]:
np.percentile(df['Price'] , 47)

5.01625

Comments:
- 75% of the stocks have a price of under 15 USD, indicating that this segment of the stock market is mostly dominated by companies with low stock price.
- 25% of the stocks have a price of under 1.53 USD, which is below the price at which some academics would consider to be penny stocks (i.e. stocks less than 5 USD)
- roughly 47% of stocks in this universe would be considered as penny stocks as they are 5 USD or less


In [13]:
df.head()


ticker       date   open  high    low  close   volume  dividends  \
0   AAME 2015-01-02  3.990  4.03  3.980   4.03  11443.0        0.0   
1   AAME 2015-01-05  3.900  4.01  3.900   4.01  13727.0        0.0   
2   AAME 2015-01-06  3.950  3.95  3.750   3.92   9743.0        0.0   
3   AAME 2015-01-07  3.899  3.92  3.870   3.92   1486.0        0.0   
4   AAME 2015-01-08  3.920  3.95  3.915   3.95   2200.0        0.0   

   closeunadj lastupdated    Price  
0        4.03  2018-06-13  4.00750  
1        4.01  2018-06-13  3.95500  
2        3.92  2018-06-13  3.89250  
3        3.92  2018-06-13  3.90225  
4        3.95  2018-06-13  3.93375

In [27]:
len(df.ticker.unique())

1948

In [98]:
tickers.head(10)

ticker
0   ABIO
1   AGTC
2    AIM
3   AIRT
4   ALSK
5    ALT
6   AVGR
7   BCDA
8   BEAT
9   BLIN

### Growth

In [70]:
 df.ticker.unique()

array(['AAME', 'AAOI', 'AAPC', ..., 'ZQKSQ', 'ZSAN', 'ZYXI'], dtype=object)

In [71]:
tickers['ticker'].to_numpy()

array(['ABIO', 'AGTC', 'AIM', ..., 'JUHL', 'OPST', 'IDAH'], dtype=object)

In [77]:
df.date.min()

Timestamp('2015-01-02 00:00:00')

In [ ]:
why is BCAR in stock_names but not price_mapping?
#i think its because the starting date is different for some stocks

In [86]:
stock_names = tickers['ticker'].to_numpy()
#store the stock ticker as stock_names

day_prices = df[df.date == df.date.min()].Price
# find the day prices at the start of our time period which is 2015-01-02

price_mapping = {n : c for n, c in zip(stock_names, day_prices)}
# Map the price of each stock to that


In [115]:
if 'BCAR' in stock_names:
  print("yes")
else:
  print("no")

yes


In [100]:
price_mapping

{'ABIO': 4.0075,
 'AGTC': 11.0175,
 'AIM': 1.77,
 'AIRT': 1.6625,
 'ALSK': 12.54,
 'ALT': 3.3425,
 'AVGR': 0.28750000000000003,
 'BCDA': 127.071,
 'BEAT': 0.13,
 'BLIN': 63.542,
 'BMTM': 4.0275,
 'CASI': 0.755,
 'CBAT': 5065.55425,
 'CDZI': 17.2375,
 'CEI': 4.49,
 'CJJD': 21.2625,
 'CLIR': 6.7749999999999995,
 'CLRB': 12.325,
 'COHN': 3.0,
 'CPHC': 20.0425,
 'CRTD': 2.42525,
 'CVV': 4.38,
 'CYCC': 8.52275,
 'CYTK': 8.6475,
 'DSS': 0.63625,
 'ENZ': 11.350000000000001,
 'FDUS': 2.09,
 'FKWL': 6.0975,
 'FORD': 0.24,
 'GV': 4.435,
 'HALL': 23.977499999999996,
 'HBIO': 316.2,
 'IBIO': 49.129999999999995,
 'IGXT': 0.11,
 'INSE': 2.69,
 'MFIN': 58.089999999999996,
 'MKTY': 0.004,
 'MLR': 30.3175,
 'MNTX': 0.08,
 'MPB': 2.3875,
 'MVBF': 25.4,
 'MVC': 1.24,
 'NANX': 16.677500000000002,
 'NBSE': 0.004,
 'NLTX': 8.3475,
 'OPTT': 1.75,
 'PEIX': 6.2425,
 'PFBX': 21.2025,
 'PRCP': 12.4305,
 'PRKR': 10.3875,
 'PTE': 9.5075,
 'PW': 3.5250000000000004,
 'RAIL': 1.06,
 'RELV': 87.01024999999998,
 'RMNI'

In [107]:
df

ticker       date    open   high     low  close    volume  dividends  \
0         AAME 2015-01-02   3.990   4.03   3.980   4.03   11443.0        0.0   
1         AAME 2015-01-05   3.900   4.01   3.900   4.01   13727.0        0.0   
2         AAME 2015-01-06   3.950   3.95   3.750   3.92    9743.0        0.0   
3         AAME 2015-01-07   3.899   3.92   3.870   3.92    1486.0        0.0   
4         AAME 2015-01-08   3.920   3.95   3.915   3.95    2200.0        0.0   
...        ...        ...     ...    ...     ...    ...       ...        ...   
2397143   ZYXI 2020-12-14  13.100  13.59  13.100  13.34  446678.0        0.0   
2397144   ZYXI 2020-12-15  13.450  13.45  13.030  13.18  372095.0        0.0   
2397145   ZYXI 2020-12-16  13.240  13.95  13.240  13.87  560581.0        0.0   
2397146   ZYXI 2020-12-17  13.940  14.29  13.701  14.25  409889.0        0.0   
2397147   ZYXI 2020-12-18  14.250  14.39  13.730  14.00  551332.0        0.0   

         closeunadj lastupdated     Price  
0              4.03  2018-06-13   4.00750  
1              4.01  2018-06-13   3.95500  
2              3.92  2018-06-13   3.89250  
3              3.92  2018-06-13   3.90225  
4              3.95  2018-06-13   3.93375  
...             ...         ...       ...  
2397143       13.34  2020-12-14  13.28250  
2397144       13.18  2020-12-15  13.27750  
2397145       13.87  2020-12-16  13.57500  
2397146       14.25  2020-12-17  14.04525  
2397147       14.00  2020-12-18  14.09250  

[2397148 rows x 11 columns]

In [109]:
df[df['ticker'] =='BCAR']

ticker       date   open   high    low  close    volume  dividends  \
229311   BCAR 2015-01-02  0.140  0.140  0.130  0.140   44080.0        0.0   
229312   BCAR 2015-01-05  0.140  0.150  0.135  0.135  101798.0        0.0   
229313   BCAR 2015-01-06  0.140  0.140  0.135  0.135   42469.0        0.0   
229314   BCAR 2015-01-07  0.140  0.150  0.140  0.140  606992.0        0.0   
229315   BCAR 2015-01-08  0.140  0.140  0.140  0.140   69800.0        0.0   
...       ...        ...    ...    ...    ...    ...       ...        ...   
229453   BCAR 2015-07-28  0.135  0.135  0.135  0.135       0.0        0.0   
229454   BCAR 2015-07-29  0.135  0.135  0.135  0.135    5000.0        0.0   
229455   BCAR 2015-07-30  0.134  0.134  0.134  0.134    1500.0        0.0   
229456   BCAR 2015-07-31  0.134  0.134  0.134  0.134    1150.0        0.0   
229457   BCAR 2015-08-03  0.134  0.136  0.134  0.134   13107.0        0.0   

        closeunadj lastupdated   Price  
229311       0.140  2018-06-13  0.1375  
229312       0.135  2018-06-13  0.1400  
229313       0.135  2018-06-13  0.1375  
229314       0.140  2018-06-13  0.1425  
229315       0.140  2018-06-13  0.1400  
...            ...         ...     ...  
229453       0.135  2018-10-16  0.1350  
229454       0.135  2018-06-13  0.1350  
229455       0.134  2018-06-13  0.1340  
229456       0.134  2018-06-13  0.1340  
229457       0.134  2018-06-13  0.1345  

[147 rows x 11 columns]

In [126]:
df.head(100)


ticker       date   open   high    low  close   volume  dividends  \
0    AAME 2015-01-02  3.990  4.030  3.980  4.030  11443.0       0.00   
1    AAME 2015-01-05  3.900  4.010  3.900  4.010  13727.0       0.00   
2    AAME 2015-01-06  3.950  3.950  3.750  3.920   9743.0       0.00   
3    AAME 2015-01-07  3.899  3.920  3.870  3.920   1486.0       0.00   
4    AAME 2015-01-08  3.920  3.950  3.915  3.950   2200.0       0.00   
5    AAME 2015-01-09  3.950  4.010  3.940  4.000   3651.0       0.00   
6    AAME 2015-01-12  4.000  4.000  4.000  4.000   1027.0       0.00   
7    AAME 2015-01-13  3.910  4.000  3.910  4.000   1311.0       0.00   
8    AAME 2015-01-14  3.890  4.000  3.890  4.000    220.0       0.00   
9    AAME 2015-01-15  3.900  4.000  3.800  3.940   8475.0       0.00   
10   AAME 2015-01-16  3.980  4.000  3.810  4.000   1150.0       0.00   
11   AAME 2015-01-20  3.821  3.990  3.821  3.990   1107.0       0.00   
12   AAME 2015-01-21  3.990  3.990  3.990  3.990     33.0       0.00   
13   AAME 2015-01-22  3.820  4.000  3.820  4.000   6880.0       0.00   
14   AAME 2015-01-23  3.850  3.880  3.850  3.880   1300.0       0.00   
15   AAME 2015-01-26  3.960  3.990  3.870  3.870   1422.0       0.00   
16   AAME 2015-01-27  3.970  3.990  3.970  3.990    700.0       0.00   
17   AAME 2015-01-28  3.903  4.000  3.903  4.000   2593.0       0.00   
18   AAME 2015-01-29  3.900  3.900  3.900  3.900    400.0       0.00   
19   AAME 2015-01-30  3.790  3.990  3.790  3.990   4236.0       0.00   
20   AAME 2015-02-02  3.970  3.990  3.970  3.990    500.0       0.00   
21   AAME 2015-02-03  3.900  3.990  3.900  3.911    524.0       0.00   
22   AAME 2015-02-04  3.990  4.000  3.970  3.970   2458.0       0.00   
23   AAME 2015-02-05  4.000  4.020  3.990  3.990   1032.0       0.00   
24   AAME 2015-02-06  3.990  3.990  3.860  3.990   3445.0       0.00   
25   AAME 2015-02-09  3.970  3.980  3.884  3.980   3872.0       0.00   
26   AAME 2015-02-10  3.980  3.980  3.980  3.980      0.0       0.00   
27   AAME 2015-02-11  3.820  3.980  3.820  3.960   5433.0       0.00   
28   AAME 2015-02-12  3.980  3.980  3.970  3.970   1100.0       0.00   
29   AAME 2015-02-13  3.970  3.980  3.830  3.960   2310.0       0.00   
30   AAME 2015-02-17  3.970  4.010  3.861  3.910   5091.0       0.00   
31   AAME 2015-02-18  3.857  3.990  3.857  3.990    799.0       0.00   
32   AAME 2015-02-19  3.910  3.960  3.830  3.900   8332.0       0.00   
33   AAME 2015-02-20  3.970  3.990  3.970  3.990    746.0       0.00   
34   AAME 2015-02-23  3.990  4.010  3.990  4.010   1825.0       0.00   
35   AAME 2015-02-24  3.940  4.030  3.940  4.020   1225.0       0.00   
36   AAME 2015-02-25  4.020  4.020  4.020  4.020     91.0       0.00   
37   AAME 2015-02-26  3.950  4.020  3.910  4.020   2385.0       0.00   
38   AAME 2015-02-27  4.020  4.030  3.926  4.000   2563.0       0.00   
39   AAME 2015-03-02  4.020  4.020  3.974  4.020   1738.0       0.00   
40   AAME 2015-03-03  4.020  4.080  3.930  4.026   4037.0       0.00   
41   AAME 2015-03-04  4.060  4.060  3.928  3.928    606.0       0.00   
42   AAME 2015-03-05  3.990  3.990  3.980  3.980   1218.0       0.00   
43   AAME 2015-03-06  4.020  4.020  3.940  3.970   4064.0       0.00   
44   AAME 2015-03-09  3.970  4.020  3.970  4.010   1426.0       0.00   
45   AAME 2015-03-10  4.020  4.020  4.010  4.010   1159.0       0.00   
46   AAME 2015-03-11  4.020  4.050  3.900  3.920   4599.0       0.00   
47   AAME 2015-03-12  4.020  4.020  3.980  3.990   4096.0       0.00   
48   AAME 2015-03-13  4.020  4.020  4.020  4.020   1269.0       0.00   
49   AAME 2015-03-16  4.020  4.020  4.020  4.020   1037.0       0.00   
50   AAME 2015-03-17  4.000  4.020  4.000  4.020   8900.0       0.00   
51   AAME 2015-03-18  4.020  4.050  4.020  4.040    971.0       0.00   
52   AAME 2015-03-19  4.000  4.040  3.980  4.030   5557.0       0.00   
53   AAME 2015-03-20  3.980  4.080  3.916  3.950   7541.0       0.00   
54   AAME 2015-03-23  3.940  4.050  3.940  

In [119]:
df.isnull().sum()

ticker         0
date           0
open           0
high           0
low            0
close          0
volume         0
dividends      0
closeunadj     0
lastupdated    0
Price          0
dtype: int64

In [116]:
if 'BCAR' in price_mapping:
  print("yes")
else:
  print("no")

no


In [104]:
base_mapping = np.array(list(map(lambda x : price_mapping[x], df['ticker'].values)))
df['Growth'] = df['Price'] / base_mapping - 1

#how do i make lambda pass if the ticker doesn't exist at the time i specify?

KeyError: 'BCAR'

## Time Series Analysis
- we want to find out the top 5 best and worst performing stocks during our chosen time period


In [14]:
sample_dates = pd.date_range(start='2015-01-02', end='2020-12-18', freq='B')


In [15]:
year_end_dates = sample_dates[sample_dates.is_year_end]
year_end_dates
# we take then end of the year for each period, except for the year 

DatetimeIndex(['2015-12-31', '2016-12-30', '2017-12-29', '2018-12-31',
               '2019-12-31'],
              dtype='datetime64[ns]', freq=None)

In [16]:
type(year_end_dates)

pandas.core.indexes.datetimes.DatetimeIndex

In [17]:
 need to add this into year_end_dates('2020-12-18')

SyntaxError: invalid syntax (<ipython-input-17-89c9475ab75f>, line 1)

In [ ]:
worst_stocks = price_volume_sorted[price_volume_sorted.date == price_volume_sorted.date.max()].sort_values('Growth').head(5)


In [ ]:
price_volume_sorte

In [ ]:
price_volume_sorted.set_index('date', inplace=True)
